File: integration-tests-java/src/main/java/io/micronaut/jimmer/java/it/config/BaseEntityDraftInterceptor.java
package io.micronaut.jimmer.java.it.config;

import io.micronaut.jimmer.java.it.entity.BaseEntity;
import io.micronaut.jimmer.java.it.entity.BaseEntityDraft;
import io.micronaut.jimmer.java.it.entity.BaseEntityProps;
import jakarta.inject.Singleton;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.Collection;
import org.babyfish.jimmer.ImmutableObjects;
import org.babyfish.jimmer.meta.TypedProp;
import org.babyfish.jimmer.sql.DraftInterceptor;
import org.jetbrains.annotations.Nullable;

@Singleton
public class BaseEntityDraftInterceptor implements DraftInterceptor<BaseEntity, BaseEntityDraft> {

    @Override
    public void beforeSave(BaseEntityDraft draft, @Nullable BaseEntity original) {
        if (!ImmutableObjects.isLoaded(draft, BaseEntityProps.MODIFIED_TIME)) {
            draft.setModifiedTime(LocalDateTime.now());
        }
        if (original == null) {
            if (!ImmutableObjects.isLoaded(draft, BaseEntityProps.CREATED_TIME)) {
                draft.setCreatedTime(LocalDateTime.now());
            }
        }
    }

    @Nullable
    @Override
    public Collection<TypedProp<BaseEntity, ?>> dependencies() {
        return Arrays.asList(BaseEntityProps.CREATED_TIME, BaseEntityProps.MODIFIED_TIME);
    }
}

File: integration-tests-kotlin/src/main/kotlin/io/micronaut/jimmer/kotlin/it/config/BaseEntityDraftInterceptor.kt
package io.micronaut.jimmer.kotlin.it.config

import io.micronaut.jimmer.kotlin.it.entity.BaseEntity
import io.micronaut.jimmer.kotlin.it.entity.BaseEntityDraft
import io.micronaut.jimmer.kotlin.it.entity.BaseEntityProps
import jakarta.inject.Singleton
import org.babyfish.jimmer.ImmutableObjects
import org.babyfish.jimmer.meta.TypedProp
import org.babyfish.jimmer.sql.DraftInterceptor
import java.time.LocalDateTime

@Singleton
class BaseEntityDraftInterceptor : DraftInterceptor<BaseEntity, BaseEntityDraft> {
    override fun beforeSave(
        draft: BaseEntityDraft,
        original: BaseEntity?,
    ) {
        if (!ImmutableObjects.isLoaded(draft, BaseEntityProps.MODIFIED_TIME)) {
            draft.modifiedTime = LocalDateTime.now()
        }
        if (original == null) {
            if (!ImmutableObjects.isLoaded(draft, BaseEntityProps.CREATED_TIME)) {
                draft.createdTime = LocalDateTime.now()
            }
        }
    }

    override fun dependencies(): Collection<TypedProp<BaseEntity, *>> = listOf(BaseEntityProps.CREATED_TIME, BaseEntityProps.MODIFIED_TIME)
}
